photon-ml/src/main/scala/com/linkedin/photon/ml/function/glm/IndividualGLMLossFunction.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function.glm
+
+import breeze.linalg._
+import org.apache.spark.broadcast.Broadcast
+
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.function._
+import com.linkedin.photon.ml.normalization.NormalizationContext
+import com.linkedin.photon.ml.optimization.{GLMOptimizationConfiguration, RegularizationType}
+
+/**
+ * This class is used to calculate the value, gradient, and Hessian of generalized linear models for individual
+ * optimization problems. The loss function of a generalized linear model can all be expressed as:
+ *
+ * L(w) = \sum_i l(z_i, y_i)
+ *
+ * with:
+ *
+ * z_i = w^T^ z_i.
+ *
+ * Different generalized linear models will have different l(z, y). The functionality of l(z, y) is provided by a
+ * [[PointwiseLossFunction]]. Since the loss function could change for different types of normalization, a normalization
+ * context object indicates which normalization strategy is used to evaluate the loss function.
+ *
+ * @param singleLossFunction A single loss function l(z, y) used for the generalized linear model
+ */
+protected[ml] class IndividualGLMLossFunction private (singleLossFunction: PointwiseLossFunction)
+  extends IndividualObjectiveFunction
+  with TwiceDiffFunction {
+
+  /**
+   * Compute the value of the function over the given data for the given model coefficients.
+   *
+   * @param input The given data over which to compute the objective value
+   * @param coefficients The model coefficients used to compute the function's value
+   * @param normalizationContext The normalization context
+   * @return The computed value of the function
+   */
+  override protected[ml] def value(
+    input: Iterable[LabeledPoint],
+    coefficients: Vector[Double],
+    normalizationContext: Broadcast[NormalizationContext]): Double =
```

Comment:
There's only one normalization context for the whole training job (for both Photon and GAME, though for GAME it's always `NoNormalization()` as normalization is disabled for GAME currently). To reduce network overhead, it should be broadcast. Even the `IndividualOptimizationProblem` objects need a broadcast normalization context, as though they themselves solve their problem local to one node, they should all use the same normalization context and they are distributed across the cluster.

Note that this may need to be revisited as part of adding normalization to GAME (coming this quarter, I believe).